[UnifiedPDF] PDFPluginBase should notify the web page of selection changes

aprotyas opened the pull request:
#### 749aa4442b2f96f7eef14849d68352302839728f
<pre>
[UnifiedPDF] PDFPluginBase should notify the web page of selection changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=268618">https://bugs.webkit.org/show_bug.cgi?id=268618</a>
<a href="https://rdar.apple.com/122179410">rdar://122179410</a>

Reviewed by Simon Fraser.

In anticipation of selection tracking support in UnifiedPDFPlugin, this
patch factors out PDFPlugin::notifySelectionChanged into a common space.

We also take the liberty to drop the PDFSelection argument to
notifySelectionChanged, given that it was not being read or passed to the
web page by the PDF plugin.

Lastly, we add a WebFrame::protectedCoreLocalFrame getter to better
abide by our smart pointer rules, where we protect both the objects on
which we call non-trivial member functions, and parameters we pass to
such functions.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(-[WKPDFLayerControllerDelegate pdfLayerController:didChangeSelection:]):
(WebKit::PDFPlugin::notifySelectionChanged): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::notifySelectionChanged):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::protectedCoreLocalFrame const):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:

Canonical link: <a href="https://commits.webkit.org/274005@main">https://commits.webkit.org/274005@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d46739f2c991aac30310fc1769aac10eda429874

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37528 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40054 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/33438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19032 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13577 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38093 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/13836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12049 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41318 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/33922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37917 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12588 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/36078 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/14041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8449 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12993 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -187,6 +187,13 @@
+{
+    if (!m_frame || !m_frame->page())
+        return;
+    m_frame->page()->didChangeSelection(*m_frame->coreLocalFrame());
```
```
m_frame->protectedPage()->didChangeSelection(*RefPtr { m_frame->coreLocalFrame() });
```
or
```
m_frame->protectedPage()->didChangeSelection(*m_frame->protectedCoreLocalFrame());
```

if you add a `protectedCoreLocalFrame()` getter that returns a RefPtr.

cdumez replied:
Per our smart pointer rules, you need to protect both:
1. objects you call a non-trivial member function on
2. parameters you pass to such functions

aprotyas replied:
Fixing shortly, thanks for the catch - didn't notice the unprotected use when I was shuffling things around.

aprotyas replied:
Fixed in [a3dd9df](https://github.com/WebKit/WebKit/commit/a3dd9dfcbfecae40efc594c4f407c43cf4fbcab9)

cdumez reviewed (commented):
LGTM but I'll let a proper PDFPlugin engineer do the formal review.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -473,7 +473,8 @@ - (void)pdfLayerController:(PDFLayerController *)pdfLayerController didChangeDis
 - (void)pdfLayerController:(PDFLayerController *)pdfLayerController didChangeSelection:(PDFSelection *)selection
 {
-    _pdfPlugin->notifySelectionChanged(selection);
+    UNUSED_PARAM(selection);
```
You may not need this. Harmless but ugly.

aprotyas replied:
Fair, it's not necessary. I'll drop it.

aprotyas replied:
Removed in [d46739f](https://github.com/WebKit/WebKit/commit/d46739f2c991aac30310fc1769aac10eda429874)

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -187,6 +187,13 @@
+{
+    if (!m_frame || !m_frame->page())
+        return;
+    m_frame->protectedPage()->didChangeSelection(*m_frame->protectedCoreLocalFrame());
```
We should think a little bit about whether it makes sense to trigger `didChangeSelection()` for inside-PDF selections, Probably fine, but other code might get confused about selection in HTML content not actually changing.

aprotyas replied:
Probably fine and also not doing much? Maybe hortont424 knows more

aprotyas committed:
[UnifiedPDF] PDFPluginBase should notify the web page of selection changes
https://bugs.webkit.org/show_bug.cgi?id=268618
rdar://122179410

Reviewed by Simon Fraser.

In anticipation of selection tracking support in UnifiedPDFPlugin, this
patch factors out PDFPlugin::notifySelectionChanged into a common space.

We also take the liberty to drop the PDFSelection argument to
notifySelectionChanged, given that it was not being read or passed to the
web page by the PDF plugin.

Lastly, we add a WebFrame::protectedCoreLocalFrame getter to better
abide by our smart pointer rules, where we protect both the objects on
which we call non-trivial member functions, and parameters we pass to
such functions.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(-[WKPDFLayerControllerDelegate pdfLayerController:didChangeSelection:]):
(WebKit::PDFPlugin::notifySelectionChanged): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::notifySelectionChanged):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::protectedCoreLocalFrame const):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:

Canonical link: https://commits.webkit.org/274005@main
